presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardCompactionManager.java

```diff
@@ -234,4 +230,87 @@ private boolean needsCompaction(ShardMetadata shard)
+        {
+            checkArgument(maxShardRows > 0, "maxShardRows must be > 0");
+
+            this.maxShardSize = requireNonNull(maxShardSize, "maxShardSize is null");
+            this.maxShardRows = maxShardRows;
+        }
+
+        public Set<OrganizationSet> createCompactionSets(Table tableInfo, Collection<ShardIndexInfo> shards)
+        {
+            Collection<Collection<ShardIndexInfo>> shardsByDaysBuckets = getShardsByDaysBuckets(tableInfo, shards);
+
+            ImmutableSet.Builder<OrganizationSet> compactionSets = ImmutableSet.builder();
+            for (Collection<ShardIndexInfo> shardInfos : shardsByDaysBuckets) {
+                compactionSets.addAll(buildCompactionSets(tableInfo, ImmutableSet.copyOf(shardInfos)));
+            }
+            return compactionSets.build();
+        }
+
+        private Set<OrganizationSet> buildCompactionSets(Table tableInfo, Set<ShardIndexInfo> shardIndexInfos)
+        {
+            long tableId = tableInfo.getTableId();
+            List<ShardIndexInfo> shards = shardIndexInfos.stream()
+                    .sorted(getShardIndexInfoComparator(tableInfo))
+                    .collect(toCollection(ArrayList::new));
+
+            long consumedBytes = 0;
+            long consumedRows = 0;
+            ImmutableSet.Builder<ShardIndexInfo> builder = ImmutableSet.builder();
+            ImmutableSet.Builder<OrganizationSet> compactionSets = ImmutableSet.builder();
+
+            for (ShardIndexInfo shard : shards) {
+                if (((consumedBytes + shard.getUncompressedSize()) > maxShardSize.toBytes()) ||
+                        (consumedRows + shard.getRowCount() > maxShardRows)) {
+                    // Finalize this compaction set, and start a new one for the rest of the shards
+                    Set<ShardIndexInfo> shardsToCompact = builder.build();
+
+                    if (shardsToCompact.size() > 1) {
+                        compactionSets.add(createOrganizationSet(tableId, shardsToCompact));
+                    }
+
+                    builder = ImmutableSet.builder();
+                    consumedBytes = 0;
+                    consumedRows = 0;
+                }
+                builder.add(shard);
+                consumedBytes += shard.getUncompressedSize();
+                consumedRows += shard.getRowCount();
+            }
+
+            // create compaction set for the remaining shards of this day
+            Set<ShardIndexInfo> shardsToCompact = builder.build();
+            if (shardsToCompact.size() > 1) {
+                compactionSets.add(createOrganizationSet(tableId, shardsToCompact));
+            }
+            return compactionSets.build();
+        }
+
+        private static Comparator<ShardIndexInfo> getShardIndexInfoComparator(Table tableInfo)
+        {
+            if (!tableInfo.getTemporalColumnId().isPresent()) {
+                return comparing(ShardIndexInfo::getUncompressedSize);
+            }
+
+            return (o1, o2) -> {
```

Comment:
I think this can be simplified
```java
return comparing(info -> info.getTemporalRange().get(),
       comparing(ShardRange::getMinTuple)
                .thenComparing(ShardRange::getMaxTuple));